File: packages/sdk/src/ensemble.ts
import { BigNumberish, ethers } from "ethers";
import { AgentData, ContractConfig, Proposal, TaskData, TaskCreationParams } from "./types";
import { ContractService } from "./services/ContractService";
import { TaskService } from "./services/TaskService";
import { AgentService } from "./services/AgentService";
import { ProposalService } from "./services/ProposalService";
import TaskRegistryABI from './abi/TaskRegistry.abi.json';
import AgentRegistryABI from './abi/AgentsRegistry.abi.json';

export class Ensemble {
  protected contractService: ContractService;
  private taskService: TaskService;
  private agentService: AgentService;
  private proposalService: ProposalService;

  constructor(config: ContractConfig, signer: ethers.Signer) {
    console.log('Config Params:', {
      network: {
        rpcUrl: config.network.rpcUrl,
        chainId: config.network.chainId,
        name: config.network.name
      },
      taskRegistryAddress: config.taskRegistryAddress,
      agentRegistryAddress: config.agentRegistryAddress
    });
    this.contractService = new ContractService(
      new ethers.JsonRpcProvider(config.network.rpcUrl),
      signer
    );

    // Initialize services
    const taskRegistry = this.contractService.createContract(
      config.taskRegistryAddress,
      TaskRegistryABI
    );
    
    const agentRegistry = this.contractService.createContract(
      config.agentRegistryAddress,
      AgentRegistryABI
    );

    this.taskService = new TaskService(taskRegistry, signer);
    this.agentService = new AgentService(agentRegistry, signer);
    this.proposalService = new ProposalService({
      projectId: 'ensemble-ai-443111',
      topicName: 'ensemble-tasks',
      taskRegistry: taskRegistry
    });
  }
  
  async start() {
    this.taskService.subscribe()
    await this.proposalService.setupSubscription(await this.agentService.getAddress());
  }

  async stop() {
    
  }

  /**
   * Creates a new task.
   * @param {TaskCreationParams} params - The parameters for task creation.
   * @returns {Promise<TaskData>} A promise that resolves to the task ID.
   */
  async createTask(params: TaskCreationParams): Promise<TaskData> {
    return this.taskService.createTask(params);
  }

  /**
   * Gets data for a specific task.
   * @param {string} taskId - The ID of the task.
   * @returns {Promise<TaskData>} A promise that resolves to the task data.
   */
  async getTaskData(taskId: string): Promise<TaskData> {
    return this.taskService.getTaskData(taskId);
  }

  /**
   * Gets tasks by owner.
   * @param {string} owner - The owner of the tasks.
   * @returns {Promise<string[]>} A promise that resolves to the task IDs.
   */
  async getTasksByOwner(owner: string): Promise<TaskData[]> {
    return this.taskService.getTasksByOwner(owner);
  }

  /**
   * Completes a task.
   * @param {BigNumberish} taskId - The ID of the task.
   * @param {string} result - The result of the task.
   * @returns {Promise<void>} A promise that resolves when the task is completed.
   */
  async completeTask(taskId: BigNumberish, result: string): Promise<void> {
    return this.taskService.completeTask(taskId, result);
  }
  
  /**
   * Registers a new agent.
   * @param {string} model - The model of the agent.
   * @param {string} prompt - The prompt for the agent.
   * @param {string[]} skills - The skills of the agent.
   * @returns {Promise<string>} A promise that resolves to the agent address.
   */
  async registerAgent(model: string, prompt: string, skills: string[]): Promise<string> {
    return this.agentService.registerAgent(model, prompt, skills);
  }

  /**
   * Gets the address of the agent.
   * @returns {Promise<string>} A promise that resolves to the agent address.
   */
  async getWalletAddress(): Promise<string> {
    return this.agentService.getAddress();
  }

  /**
   * Gets data for a specific agent.
   * @param {string} agentAddress - The address of the agent.
   * @returns {Promise<AgentData>} A promise that resolves to the agent data.
   */
  async getAgentData(agentId: string): Promise<AgentData> {
    return this.agentService.getAgentData(agentId);
  }

  /**
   * Checks if an agent is registered.
   * @param {string} agentAddress - The address of the agent.
   * @returns {Promise<boolean>} A promise that resolves to a boolean indicating if the agent is registered.
   */
  async isAgentRegistered(agentId: string): Promise<boolean> {
    return this.agentService.isAgentRegistered(agentId);
  }


  /**
   * Sends a proposal for a task.
   * @param {string} taskId - The ID of the task.
   * @param {BigNumberish} price - The price of the proposal.
   * @returns {Promise<void>} A promise that resolves when the proposal is sent.
   */
  async sendProposal(taskId: string, price: BigNumberish): Promise<void> {
    return this.proposalService.sendProposal(taskId, await this.agentService.getAddress(), price);
  }

  /**
   * Gets proposals for a task.
   * @param {string} taskId - The ID of the task.
   * @returns {Promise<string[]>} A promise that resolves to an array of proposal IDs.
   */
  async getProposals(taskId: string): Promise<string[]> {
    return this.proposalService.getProposals(taskId);
  }

  /**
   * Approves a proposal for a task.
   * @param {BigNumberish} taskId - The ID of the task.
   * @param {Object} proposal - The proposal object.
   * @param {BigNumberish} proposal.id - The ID of the proposal.
   * @param {BigNumberish} proposal.price - The price of the proposal.
   * @param {BigNumberish} proposal.taskId - The ID of the task.
   * @param {string} proposal.agent - The agent address.
   * @returns {Promise<void>} A promise that resolves when the proposal is approved.
   */
  async approveProposal(taskId: BigNumberish, proposal: Proposal): Promise<void> {
    return this.proposalService.approveProposal(taskId, proposal);
  }

  async setOnNewTaskListener(listener: (task: TaskData) => void) {
    return this.taskService.setOnNewTaskListener(listener);
  }

  async setOnNewProposalListener(listener: (proposal: Proposal) => void) {
    return this.proposalService.setOnNewProposalListener(listener);
  }
} 

File: packages/python-sdk/src/ensemble.py

from web3 import Web3
from eth_typing import Address
from .services.agent_service import AgentService
from .services.task_service import TaskService
from .services.proposal_service import ProposalService
from .services.contract_service import ContractService
from .types import ContractConfig, Proposal
from typing import Callable, Optional

class Ensemble:
    def __init__(self, config: ContractConfig):
        self.contract_service = ContractService(
            Web3(Web3.HTTPProvider(config.network.rpc_url)),
            None  # Account will be set in connect()
        )
        
        # Load contract ABIs
        with open('src/abi/AgentsRegistry.abi.json') as f:
            agents_registry_abi = f.read()
        with open('src/abi/TaskRegistry.abi.json') as f:
            task_registry_abi = f.read()
            
        self.agent_registry = self.contract_service.create_contract(
            config.agent_registry_address,
            agents_registry_abi
        )
        self.task_registry = self.contract_service.create_contract(
            config.task_registry_address,
            task_registry_abi
        )
        
        self.config = config
        
    async def connect(self, private_key: str):
        account = self.contract_service.provider.eth.account.from_key(private_key)
        self.contract_service.account = account
        
        # Initialize services
        self.agent_service = AgentService(self.agent_registry, account)
        self.task_service = TaskService(self.task_registry, account)
        self.proposal_service = ProposalService(
            project_id='ensemble-ai-443111',
            topic_name='ensemble-tasks',
            task_registry=self.task_registry
        )
        
        await self.contract_service.validate_network(self.config.network.chain_id)
        return self
        
    async def start(self):
        user_id = await self.agent_service.get_address()
        await self.proposal_service.setup_subscription(user_id)
        
    async def stop(self):
        if self.proposal_service.subscription:
            self.proposal_service.subscription.close()
            
    # Agent Management
    async def register_agent(self, model: str, prompt: str, skills: list[str]) -> str:
        return await self.agent_service.register_agent(model, prompt, skills)
        
    async def get_agent_data(self, agent_id: str):
        return await self.agent_service.get_agent_data(agent_id)
        
    async def is_agent_registered(self, agent_id: str) -> bool:
        return await self.agent_service.is_agent_registered(agent_id)
        
    # Task Management
    async def create_task(self, params: dict) -> int:
        return await self.task_service.create_task(params)
        
    async def get_task_data(self, task_id: str) -> dict:
        return await self.task_service.get_task_data(task_id)
        
    async def get_tasks_by_owner(self, owner: str) -> list:
        return await self.task_service.get_tasks_by_owner(owner)
        
    async def complete_task(self, task_id: int, result: str) -> None:
        await self.task_service.complete_task(task_id, result)
        
    # Proposal Management
    async def send_proposal(self, task_id: str, agent_id: str, price: int) -> None:
        await self.proposal_service.send_proposal(task_id, agent_id, price)
        
    async def get_proposals(self, task_id: str) -> list[str]:
        return await self.proposal_service.get_proposals(task_id)
        
    async def approve_proposal(self, task_id: int, proposal: Proposal) -> None:
        await self.proposal_service.approve_proposal(task_id, proposal)
        
    def set_on_new_proposal_listener(self, listener: Callable[[Proposal], None]):
        self.proposal_service.set_on_new_proposal_listener(listener)
        
    async def get_wallet_address(self) -> str:
        return await self.agent_service.get_address()
